core/common/src/annotation/TextBlock.ts

```diff
@@ -0,0 +1,301 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Annotation
+ */
+
+import { TextStyle, TextStyleSettingsProps } from "./TextStyle";
+
+export interface ApplyTextStyleOptions {
+  preserveOverrides?: boolean;
+  dontPropagate?: boolean;
+}
+
+export interface TextBlockComponentProps {
+  styleName: string;
+  styleOverrides?: TextStyleSettingsProps;
+}
+
+export interface TextBlockStringifyOptions {
+  /** Default: " " */
+  paragraphBreak?: string;
+  /** Default: " " */
+  lineBreak?: string;
+  /** Default: "/" */
+  fractionSeparator?: string;
+}
+
+export abstract class TextBlockComponent {
+  private _styleName: string;
+  public styleOverrides: TextStyleSettingsProps;
+
+  /** @internal */
+  protected constructor(props: TextBlockComponentProps) {
+    this._styleName = props.styleName;
+    this.styleOverrides = { ...props.styleOverrides };
+  }
+
+  public get styleName(): string {
+    return this._styleName;
+  }
+
+  public set styleName(styleName: string) {
+    this.applyStyle(styleName);
+  }
+
+  public applyStyle(styleName: string, options?: ApplyTextStyleOptions): void {
+    this._styleName = styleName;
+
+    if (!(options?.preserveOverrides)) {
+      this.clearOverrides();
+    }
+  }
+
+  public clearOverrides(): void {
+    this.styleOverrides = { };
+  }
+
+  public get overridesStyle(): boolean {
+    return Object.keys(this.styleOverrides).length > 0;
+  }
+
+  public createEffectiveStyle(baseStyle: TextStyle): TextStyle {
+    return this.overridesStyle ? baseStyle.clone(this.styleOverrides) : baseStyle;
+  }
+
+  public abstract clone(): TextBlockComponent;
+
+  public abstract stringify(options?: TextBlockStringifyOptions): string;
+  
+  public toJSON(): TextBlockComponentProps {
+    return {
+      styleName: this.styleName,
+      styleOverrides: { ...this.styleOverrides },
+    }
+  }
+}
+  
+export type Run = TextRun | FractionRun | LineBreakRun;
+export type RunProps = TextRunProps | FractionRunProps | LineBreakRunProps;
+
+export namespace Run {
+  export function fromJSON(props: RunProps): Run {
+    switch (props.type) {
+      case "text": return TextRun.create(props);
+      case "fraction": return FractionRun.create(props);
+      case "linebreak": return LineBreakRun.create(props);
+    }
+  }
+}
+
+export type TextRunShiftMode = "subscript" | "superscript" | "baseline";
```

Comment:
There's probably a better name than "shift mode" that eludes me.

Reply:
I can't think of a unifying term off the top of my head either. One variation might be `BaselineShift = "none" | "subscript" | "superscript";`.

Reply:
vertical alignment?